WIP: receive payjoin

arminsabouri opened the pull request:
Recieve side still needs work. Plus need some cool ass UI

DanGould commented on Chaincase.Common/Contracts/ITorManager.cs:
```diff
@@ -17,5 +17,9 @@ public interface ITorManager
+
+        string CreateHiddenServiceAsync();
+
+        void DestroyHiddenServiceAsync(string serviceId);
```
We didn't think much about managing hidden service identifying keys on first pass. `DestroyHiddenServiceAsync` removes the ephemeral service if I remember correctly.

DanGould commented on Chaincase.Common/Global.cs:
```diff
@@ -229,6 +232,11 @@ public async Task InitializeNoWalletAsync(CancellationToken cancellationToken =
 
+                if (!_P2EPServer.HiddenServiceIsOn) {
+                    var cts = new CancellationToken();
+                    await _P2EPServer.StartAsync(cts);
```
Why the conditional? The synchronizer manages it's own lifecycle so if you button mash it doesn't cause problems. I think P2EPServer's lifecycle should be self contained and not reliant on external state checks

DanGould commented on Chaincase.Common/Global.cs:
```diff
@@ -480,6 +488,20 @@ public async Task OnSleeping()
+                        //    await Task.Delay(1_000); //
+                        //    if (KeepHiddenServiceAlive.IsCancellationRequested || !P2EPServer.HiddenServiceIsOn) return;
+                        //}
+                        await _P2EPServer.StopAsync(SleepCts.Token);
```
same comment as above. P2EPServer's lifecycle ought to be self contained and non reliant on external control conditions

DanGould commented on Chaincase.Common/Services/P2EPRequestHandler.cs:
```diff
@@ -0,0 +1,78 @@
+
+namespace Chaincase.Common.Services
+{
+	public class P2EPRequestHandler
```
Called a `Controller` e.g. `P2EPController` in ASP.Net parlance

DanGould commented on Chaincase.Common/Services/P2EPServer.cs:
```diff
@@ -0,0 +1,130 @@
+							var output = response.OutputStream;
+							var buffer = Encoding.UTF8.GetBytes(result);
+							await output.WriteAsync(buffer, 0, buffer.Length, stoppingToken).ConfigureAwait(false);
+							await output.FlushAsync(stoppingToken).ConfigureAwait(false);
```
5-deep nesting smells like 🍑💨

DanGould commented on Chaincase.UI/Components/Status.razor:
```diff
@@ -12,6 +12,9 @@
     </IonItem>
+    <IonItem>
+        <IonLabel>Tor Hidden Service Status: @TorHiddenServiceStatus()</IonLabel>
+    </IonItem>
```
Because the server is fragile on iOS and requires the app to be open for success, perhaps a colorful full-screen warning that presents the QR is more appropriate. We must also encourage use of iOS [share extensions](https://developer.apple.com/design/human-interface-guidelines/ios/extensions/sharing-and-actions/) rather than the app switcher or else Tor gets killed.

DanGould commented on Chaincase.UI/ViewModels/ReceiveViewModel.cs:
```diff
@@ -36,6 +47,45 @@ public void InitNextReceiveKey()
+        {
+            var cts = new CancellationToken();
+            _P2EPServer.StartAsync(cts);
+            _P2EPServer.Password = password;
```
field may be better set from the constructor so the consumer doesn't need to know this bit of _P2EPServer nternals

DanGould commented on Chaincase.UI/ViewModels/ReceiveViewModel.cs:
```diff
@@ -36,6 +47,45 @@ public void InitNextReceiveKey()
+            string walletFilePath = Path.Combine(_walletManager.WalletDirectories.WalletsDir, $"{_network}.json");
+            try
+            {
+                await Task.Run(() => KeyManager.FromFile(walletFilePath).GetMasterExtKey(password ?? ""));
```
dam we have to know about _walletManager & KeyManager here? looks like _wM should just expose a utility

DanGould commented on Chaincase.UI/ViewModels/StatusViewModel.cs:
```diff
@@ -46,14 +47,15 @@ public class StatusViewModel : ReactiveObject
         private StatusSet ActiveStatuses { get; }
 
-        public StatusViewModel(Global global, ChaincaseWalletManager walletManager, Config config, UiConfig uiConfig, BitcoinStore bitcoinStore, ChaincaseSynchronizer synchronizer)
+        public StatusViewModel(Global global, ChaincaseWalletManager walletManager, Config config, UiConfig uiConfig, BitcoinStore bitcoinStore, ChaincaseSynchronizer synchronizer, P2EPServer P2EPServer)
```
As stated earlier in the review, flashing screen or dep from the main StatusViewModel? Why?

DanGould commented on Chaincase.iOS/Services/iOSTorManager.cs:
```diff
@@ -290,5 +292,41 @@ private void CancelInitRetry()
             initRetry?.Cancel();
             initRetry = null;
         }
+
```
The next release of github.com/iCepa/Tor.framework will include official support for hosting v3 hidden services

https://github.com/iCepa/Tor.framework/issues/33#issuecomment-1046838630

DanGould commented:
I think that PR could effectively be separated and reviewed as hidden service, PayJoin Controller on top of that, and UI on top of both.

DanGould committed:
reference Xamarin.iOS

DanGould committed:
control hidden service

DanGould commented on Chaincase.Common/Services/P2EPServer.cs:
```diff
@@ -0,0 +1,130 @@
+					cancellationToken.ThrowIfCancellationRequested();
+				}
+			}
+			return await getHttpContextTask.ConfigureAwait(false);
```
Where did this cancellation code come from @armins88 ?

DanGould committed:
sketch P2EPServer

DanGould committed:
add PayJoinExtensions to DI

DanGould committed:
[draft] handle P2EPRequests
the psbt logic should be removed from  history

DanGould commented:
See the hidden service and P2EPServer portions refactored. DesktopTorManager throws on hidden service controls, so controls must be implemented or otherwise handled in the UI.

Server lifecycle, UI, and P2EPRequestHandler remain. The RequestHandler ought to consider #259 and use BtcPayServer.BIP78.Receiver. Separate derivation of and authorization to use key material dependencies from BIP78 psbt construction & negotiation will let us unit test each component.

DanGould committed:
bump NBitcoin v6

DanGould committed:
handle PayjoinReceiverError

DanGould commented:
Oops this pr lost some history. I thought github would keep @armins88's history for reference while we worked on the substeps when I force pushed. Armin, can you push your original changes to another branch for reference? Otherwise we can just look at onion78.

https://github.com/btcpayserver/BTCPayServer.BIP78 It does not seem like the best choice for receive after a deep dive. There's also no reference implementation. The library is not yet used in BtcPayServer, and the abstractions are a bit too opinionated about the receiver being what the spec describes as a "non-interactive" (payment-processor like) endpoint. Our own implementation, even based on that lib, can help us take the lead by defining the sdk.

DanGould committed:
[draft] handle PayJoin with BIP78.Receiver lib

DanGould committed:
swap P2EPRequestHandler with PayJoinReceiverWallet
ReceiverWallet is based on BtcPayServer.BIP78.Receiver

DanGould commented:
Our latest claim to fame is being the first reference implementation of https://github.com/btcpayserver/BTCPayServer.BIP78/tree/master/BIP78.Receiver. Of course being the first mean I found the missing pieces. The lib is missing proper spec-defined http responses, The way the `Context` object gets passed around isn't as functional/declarative/static/testable as I'd like it to be yet. And this first go didn't decouple the receiver from Chaincase's KeyManager yet either.

Password-in-memory ought to be swapped out in favor of spendable coins. Maybe our `PayJoinReceiverWallet` Initiate should include spendable coins along with the proposal context. Those spendableCoins can come from a `ChaincaseWalletManager` helper function. `PayJoinReceiverWallet` must be decoupled from Chaincase's wallet implementation so it can be SDK-ified. Those spendable coins MUST be disposed from memory when the receiver chooses to shut down, too.

In addition, UI remains

DanGould committed:
arm PayJoin from UI

DanGould committed:
upgrade to NBitcoin v6.0.19

DanGould commented:
BtcPayServer.BIP78 relies on NBitcoin > v6 so that has been added. It will be merged via another pull request

DanGould committed:
preview csharp v9.0

DanGould committed:
avoid heuristics in payjoin coin selection
this coin selection avoids ANY unecessary input heuristic for now. We can revisit this in a second receiver wallet version that can see coin unlinkability measures
